### PR TITLE
use binary gcd algorithm

### PR DESCRIPTION
**Pull request by simon-auch**
as the title says this commit implements the binary gcd algorithm (see wikipedia for more).
On my system this algorithm runs significantly faster (around 3/5 of the time for u32 and about half the time for u128) than the euclid version.

Sadly I have no arm device to test performance on, but as long as arm has a way to count leading zeros it should be fine.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -21,21 +21,24 @@ pub trait Gcd {
...
-                    a = b;
-                    b = r;
-                }
-
-                a
+            fn gcd(self, mut v: Self) -> Self
+            {
+                let mut u = self;
+                if u == 0{ return v}
+                if v == 0{ return u}
```
Comment (hellow554) [inline src/lib.rs]: This is badly formatted. Can you please use `rustfmt`?

File: src/lib.rs [hunk 2]
```diff
@@ -21,21 +21,24 @@ pub trait Gcd {
...
+                u = u>>(u.trailing_zeros());
+                while {
+                    v = v>>(v.trailing_zeros());
+                    if u > v {
+                        core::mem::swap(&mut v, &mut u);
+                    }
+                    v -= u; // Here v >= u.
+                    v!=0
+                } {}
+                return u << shift;
```
Comment (hellow554) [inline src/lib.rs]: `return` as a last statement is proscribed. Use `clippy` for linting your code

**Comment by simon-auch** [comment]
Done

**Comment by hellow554** [comment]
@raidwas could use please use `rustfmt` on your code in total? There are more than just the two mentioned places. Also the `while` loop looks weird. Could you use `loop` instead?

**Comment by frewsxcv** [comment]
i just added a benchmark in https://github.com/frewsxcv/rust-gcd/commit/a064ce724cc91e98c2b5378770f2a99f1e048673, which can be run with `cargo bench`. here's the results after running this on my machine (2015 macbook pro):

Screen Shot 2020-01-19 at 12 12 53 AM

so at least for me your changes improve performance on u64 and u128, but not other integer widths. maybe we should just offer both functions and allow users to chose?

**Comment by simon-auch** [comment]
I made both functions available from the trait.
To be fair, I should probably have also said what my benchmarks looked like, so I changed your benchmark code to reflect this. I benchmarked the functions over the full u8 range, and my results are still the same as before, the binary version always beeing at least as fast as the euclidian version (on average over all inputs).
The results from my benchmark run are as follows:
https://gistpreview.github.io/?97b32921e5d85302521fe6ebc25e7c65/report.html
TL;DR:

| type | binary | euclid |
| ----- | -------- | -------  |
| u8   | 7.1      | 9.6      |
| u16 | 7.0      | 9.2      |
| u32 | 6.4      | 8.9      |
| u64 | 6.6      | 13.4    |
|u128|14.3|28.3|

(all times in ms)
This benchmark of course isn't perfect either, since even for the bigger types only u8 ranges are tested, which is kind of unrealistic. Maybe it would be good to create random values in the full range of the type and benchmark with those?

**Comment by simon-auch** [comment]
When using pseudo random values the benchmarks get much more interesting:
https://gistpreview.github.io/?14c6c019d51a2615fd58a8e5047a1fbe/report_pseudo_random.html
TL;DR:


type | binary | euclid
-- | -- | --
u8 | 102us | 92us
u16 | 200us | 298us
u32 | 347us | 624us
u64 | 656us | 1.7ms
u128 | 3.4ms | 6.8ms

Compared to before we see much bigger differences in the times (as expected for bigger values), except for u8 the binary version is still faster on average (at least for this particular choice of seed and number of samples, which is also the only combination I tested, so no guarantees for the numbers above).

In the end I think providing both methods in the trait and also having the `gcd`  function for simplicity and backwards compatibility seems like a reasonable choice for me. Which function to use for the default is really up to you :)

File: src/lib.rs [hunk 3]
```diff
@@ -16,12 +16,41 @@ pub trait Gcd {
...
+                    return u;
+                }
+                let shift = (u | v).trailing_zeros();
+                u >>= shift;
+                v >>= shift;
+                u >>= u.trailing_zeros();
+                loop {
+                    v = v >> (v.trailing_zeros());
+                    if u > v {
+                        core::mem::swap(&mut v, &mut u);
```
Comment (hellow554) [inline src/lib.rs]: `mem::swap` is very expensive for two integer types. I used the [xor swap algorithm](https://en.wikipedia.org/wiki/XOR_swap_algorithm) as an alternative and the benchmark improved by 2-8%.

File: src/lib.rs [hunk 4]
```diff
@@ -33,9 +62,9 @@ macro_rules! gcd_impl {
                     let r = a % b;
                     a = b;
                     b = r;
-                }
+		}
 
-                a
+		a
```
Comment (hellow554) [inline src/lib.rs]: You have tabs inserted here

File: src/lib.rs (hunk 3 above)
Comment (simon-auch) [reply src/lib.rs]: I actually thought about doing that too, but decided to trust the compiler to be smart, obviously didn't work out ^^

File: src/lib.rs (hunk 4 above)
Comment (simon-auch) [reply src/lib.rs]: shouldn't `cargo fmt` have fixed that? Or is it because its inside a macro it doesn't work?

Comment (hellow554) [reply src/lib.rs]: IIRC `cargo fmt` does not touch macros at all

File: src/lib.rs (hunk 3 above)
Comment (simon-auch) [reply src/lib.rs]: At least on my system this seems to make no difference (only 3% improvement only for u8, no measurable improvement for the rest).
This might be because the old code gets optimized to two conditional moves, which on modern processors isn't slow at all.
I would still include the change since on other platforms things might be different and the xor swap shouldn't be slower (even if it uses one instruction more).

**Comment by simon-auch** [comment]
@frewsxcv maybe you could run the new benchmarks and see if the results changed?

File: src/lib.rs [hunk 5]
```diff
@@ -16,12 +16,44 @@ pub trait Gcd {
...
+                }
+                if v == 0 {
+                    return u;
+                }
+                let shift = (u | v).trailing_zeros();
+                u >>= shift;
+                v >>= shift;
+                u >>= u.trailing_zeros();
+                loop {
+                    v = v >> (v.trailing_zeros());
```
Comment (oherrala) [inline src/lib.rs:42]: This could use the same format as the above call for `u`. Just for consistency.

```suggestion
                    v >>= v.trailing_zeros();
```

**Review by frewsxcv — APPROVED** [review]
this is great, thanks so much @raidwas !
